In [31]:
from pyspark.sql import SparkSession
from pymongo import MongoClient
import pandas as pd

# Iniciar la sesión de Spark
spark = SparkSession.builder.appName("Clustering_Alojamientos").getOrCreate()

# Conexión a tu MongoDB Atlas
client = MongoClient("mongodb+srv://lucascheuque_db_user:27032005@cluster0.tjvu2a3.mongodb.net/")
data = list(client.proyecto_bigdata.alojamientos.find({}, {"_id": 0}))

# Crear el DataFrame de Spark
df_cluster = spark.createDataFrame(pd.DataFrame(data))

print(f"✅ ¡Datos cargados desde MongoDB! Total registros: {df_cluster.count()}")

✅ ¡Datos cargados desde MongoDB! Total registros: 4754


In [33]:
import pyspark.sql.functions as F

# 1. Calculamos la columna relacion_calidad_precio como en tu proyecto
df_cluster = df_cluster.withColumn(
    "relacion_calidad_precio", 
    (F.col("puntuacion") / F.col("precio_noche")) * 10000
)

# 2. Eliminamos las filas con NaN/Nulos para que no falle el VectorAssembler
df_cluster = df_cluster.na.drop(subset=["precio_noche", "puntuacion", "relacion_calidad_precio"])

print(" Columna calculada y filas con nulos eliminadas con éxito.")
df_cluster.select("nombre_hotel", "precio_noche", "puntuacion", "relacion_calidad_precio").show(5)

 Columna calculada y filas con nulos eliminadas con éxito.
+--------------------+------------+----------+-----------------------+
|        nombre_hotel|precio_noche|puntuacion|relacion_calidad_precio|
+--------------------+------------+----------+-----------------------+
| NH Iquique Pacifico|    240675.0|       8.6|    0.35732834735639346|
|Hotel Terrado Cav...|    150321.0|       8.3|     0.5521517286340565|
|Holiday Inn Expre...|    165750.0|       8.9|     0.5369532428355958|
|Terrado Arturo Pr...|    102675.0|       8.6|     0.8375943511078645|
|Departamento cént...|     90000.0|      10.0|     1.1111111111111112|
+--------------------+------------+----------+-----------------------+
only showing top 5 rows



In [35]:
from pyspark.ml.feature import VectorAssembler

# Agrupamos las tres variables en el vector llamado 'features'
assembler = VectorAssembler(
    inputCols=[
        "precio_noche",
        "puntuacion",
        "relacion_calidad_precio"
    ],
    outputCol="features"
)

df_features = assembler.transform(df_cluster)
print(" VectorAssembler ejecutado correctamente. Columna 'features' creada sin errores.")

 VectorAssembler ejecutado correctamente. Columna 'features' creada sin errores.


In [37]:
from pyspark.ml.clustering import KMeans

# Configura el modelo K-Means con 3 grupos y la semilla original
kmeans = KMeans(
    k=3,
    seed=1,
    featuresCol="features"
)

modelo = kmeans.fit(df_features)
print(" Modelo K-Means entrenado correctamente.")

 Modelo K-Means entrenado correctamente.


In [38]:
df_resultado = modelo.transform(df_features)

df_resultado.select(
    "nombre_hotel",
    "precio_noche",
    "puntuacion",
    "relacion_calidad_precio",
    "prediction"
).show(20, truncate=False)

+---------------------------------------------+------------+----------+-----------------------+----------+
|nombre_hotel                                 |precio_noche|puntuacion|relacion_calidad_precio|prediction|
+---------------------------------------------+------------+----------+-----------------------+----------+
|NH Iquique Pacifico                          |240675.0    |8.6       |0.35732834735639346    |0         |
|Hotel Terrado Cavancha                       |150321.0    |8.3       |0.5521517286340565     |0         |
|Holiday Inn Express - Iquique by IHG         |165750.0    |8.9       |0.5369532428355958     |0         |
|Terrado Arturo Prat Iquique                  |102675.0    |8.6       |0.8375943511078645     |0         |
|Departamento céntrico y cómodo               |90000.0     |10.0      |1.1111111111111112     |0         |
|Playa Hotel - Cavancha                       |109520.0    |8.8       |0.8035062089116144     |0         |
|Departamentos Alpro Cavancha Vista a

In [39]:
df_resultado.groupBy("prediction").count().show()

+----------+-----+
|prediction|count|
+----------+-----+
|         1|    8|
|         2|  456|
|         0| 3564|
+----------+-----+



In [40]:
from pyspark.sql.functions import avg

df_resultado.groupBy("prediction").agg(
    avg("precio_noche").alias("precio_promedio"),
    avg("puntuacion").alias("nota_promedio"),
    avg("relacion_calidad_precio").alias("conveniencia_promedio")
).show(truncate=False)

+----------+------------------+-----------------+---------------------+
|prediction|precio_promedio   |nota_promedio    |conveniencia_promedio|
+----------+------------------+-----------------+---------------------+
|1         |3021306.375       |7.1805           |0.024938174701251307 |
|2         |473553.82236842107|6.425521929824562|0.1579126669383086   |
|0         |106675.64842873176|7.072393378226713|0.9038675830881018   |
+----------+------------------+-----------------+---------------------+



Análisis de clústeres

Después de aplicar el algoritmo K-Means para dividir los alojamientos en tres grupos, pudimos identificar tres segmentos de mercado muy claros según el precio, la nota y la conveniencia de cada hotel.

El primer grupo, que corresponde al Clúster 0, concentra los hoteles que son económicos pero que tienen muy buena puntuación. El precio promedio por noche es de unos 106 mil pesos y la nota promedio supera el 7. Esto hace que sea el grupo con la mejor relación calidad-precio, representando las opciones más convenientes y atractivas para la mayoría de los usuarios que buscan cuidar su presupuesto sin sacrificar calidad.

El segundo grupo, que es el Clúster 1, corresponde a los alojamientos de lujo exclusivo. Aquí el precio promedio se dispara a más de 3 millones de pesos por noche. Aunque tienen la calificación más alta de todos con un 7.18, la relación de conveniencia da casi cero porque el costo es extremadamente elevado. Este grupo representa un turismo muy exclusivo.

Por último, el Clúster 2 agrupa a los hoteles de rango medio-alto, con un precio promedio de unos 473 mil pesos por noche. Lo llamativo de este grupo es que tiene la peor nota promedio, llegando apenas a un 6.42. Esto nos muestra que hay una desconexión en este segmento, ya que los clientes están pagando un precio alto pero la experiencia que reciben no está cumpliendo con sus expectativas.

Como conclusión para el negocio, el Clúster 0 es el más importante para generar volumen de ventas y atraer clientes, mientras que los hoteles del Clúster 2 necesitan revisar de forma urgente sus tarifas o mejorar sus servicios, porque los usuarios están quedando disconformes con lo que pagan